### PR TITLE
Added quality declaration draft

### PR DESCRIPTION
**Pull request by ahcorde**
This is a quality declaration for ament_index_python as described in the proposed REP-2004 (see ros-infrastructure/rep#218).

This declaration is aspirational, in that it states things which aren't currently true, but we plan to do soon so we can get it up to the highest level.

Checklist in #140 can be checked first to sort out missing parts of the declaration.

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

**Comment by ahcorde** [comment]
Missing `Documentation policy for change requests` for QD 1

File: QUALITY_DECLARATION.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+The package `class_loader` claims to be in the **Quality Level 1** category.
+
+Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 1 in REP-2004](https://www.ros.org/reps/rep-2004.html).
+
+## Version Policy [1]
+
+### Version Scheme [1.i]
+
+`class_loader` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#versioning), and is at or above a stable version. **The current version is 1.4.0**.
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: > **The current version is 1.4.0**.

Since versions change quite a bit, I would just change this to suggest where to find the current version.

Also it seems like this package just named itself 1.4 because ROS 1 uses 0.4. Does 1.x actually imply API stability?

File: QUALITY_DECLARATION.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+`class_loader` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change_control_process).
+
+### Change Requests [2.i]
+
+All changes occur through a pull request
+
+### Contributor Origin [2.ii]
+
+Must have confirmation of contributor origin (e.g. DCO, CLA, etc.).
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: This should be rephrased to state what DCO process it follows, not just repeat the info in rep-2004.

File: QUALITY_DECLARATION.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+### Peer Review Policy [2.iii]
+
+All pull request have two peer reviews
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: It looks like this repo only requires 1 peer review currently. Are there plans to change it?

File: QUALITY_DECLARATION.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+### Continuous Integration [2.iv]
+
+all pull request must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)
```
Comment (brawner) [inline QUALITY_DECLARATION.md]:
```suggestion
All pull requests must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)
```

File: QUALITY_DECLARATION.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+### Documentation Policy [2.v]
+
+All pull request must resolve related documentation changes before merging
```
Comment (brawner) [inline QUALITY_DECLARATION.md]:
```suggestion
All pull requests must resolve related documentation changes before merging
```

File: QUALITY_DECLARATION.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+## Documentation [3]
+
+### Feature Documentation [3.i]
+
+`class_loader` has a [feature list](http://wiki.ros.org/class_loader) and each item in the list links to the corresponding feature documentation.
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: This is the ROS 1 feature list, we need one for ROS 2. Or at the very least, state something to the effect of "There is no current ROS 2 feature list, but the feature list for ROS 1 can be found at..."

File: QUALITY_DECLARATION.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+There is documentation for all of the features, and new features require documentation before being added.
+
+### Public API Documentation [3.ii]
+
+`class_loader` has embedded API documentation and it is generated using doxygen and is [hosted](http://wiki.ros.org/class_loader) along side the feature documentation.
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: I don't think this should link to ros1 docs, and ROS 2 API docs aren't currently being generated for class_loader

https://github.com/ros2/docs.ros2.org/blob/a5ed03490741da161598f3b639d47eb8f361984e/Makefile#L4-L14

File: QUALITY_DECLARATION.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+The tests aim to cover both typical usage and corner cases, but are quantified by contributing to code coverage.
+
+### Coverage [4.iii]
+
+`class_loader` follows the recommendations for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#coverage), and opts to use branch coverage instead of line coverage.
+
+This includes:
+
+- tracking and reporting branch coverage statistics
+- achieving and maintaining branch coverage at or above 95%
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: I think the current coverage is 90-ish % yah?

File: QUALITY_DECLARATION.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+- no lines are manually skipped in coverage calculations
+
+Changes are required to make a best effort to keep or increase coverage before being accepted, but decreases are allowed if properly justified and accepted by reviewers.
+
+Current coverage statistics can be viewed here:
+
+![](https://codecov.io/gh/ahcorde/class_loader/branch/ahcorde_code_coverage/graphs/sunburst.svg)
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: This image isn't too helpful. Maybe you could run a ci_linux-coverage job and link that instead.

File: QUALITY_DECLARATION.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+### Performance [4.iv]
+
+`class_loader` follows the recommendations for performance testing of C++ code in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#performance-testing), and opts to do performance analysis on each release rather than each change.
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: Where are the performance tests? Just state the things that are true currently and provide links with supporting evidence.

File: QUALITY_DECLARATION.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+### Linters and Static Analysis [4.v]
+
+`class_loader` uses and passes all the standard linters and static analysis tools for a C++ package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis).
+
+## Dependencies [5]
+
+### Direct Runtime ROS Dependencies [5.i]
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: This package does depend on rcpputils

File: QUALITY_DECLARATION.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+ - console_bridge_vendor
+ - libconsole-bridge-dev
+
+It has one "buildtool" dependency, which do not affect the resulting quality of the package, because they do not contribute to the public library API.
+It also has several test dependencies, which do not affect the resulting quality of the package, because they are only used to build and run the test code.
+
+## Platform Support [6]
+
+`class_loader` supports all of the tier 1 platforms as described in [REP-2000](https://www.ros.org/reps/rep-2000.html#support-tiers), and tests each change against all of them.
```
Comment (brawner) [inline QUALITY_DECLARATION.md:144]: Add links to latest CI results. and remove TODO. I don't think we need to say anything about non-tier 1 unless we have ci tests for them.

File: QUALITY_DECLARATION.md [hunk 13]
```diff
@@ -0,0 +1,135 @@
+This document is a declaration of software quality for the `class_loader` package, based on the guidelines in [REP-2004](https://www.ros.org/reps/rep-2004.html).
+
+# `class_loader` Quality Declaration
+
+The package `class_loader` claims to be in the **Quality Level 1** category.
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: I'm not sure this quite makes it yet to QL 1 given some missing chunks below.

File: QUALITY_DECLARATION.md [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+### Direct Runtime ROS Dependencies [5.i]
+
+### Optional Direct Runtime ROS Dependencies [5.ii]
+
+### Direct Runtime non-ROS Dependency [5.iii]
+
+`class_loader` has run-time and build-time dependencies that need to be considered for this declaration.
+
+ - console_bridge_vendor
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: console_bridge_vendor is technically a ROS dep yah?

**Review by brawner — COMMENTED** [review]
I'm not sure this one is quite yet QL 1 since it's missing performance tests, and it's coverage is not yet 95%. Also, be sure to state things as they are and not how we wish them to be.

File: QUALITY_DECLARATION.md (hunk 1 above)
Comment (ahcorde) [reply QUALITY_DECLARATION.md]: Not sure about this. I assumed that the current version in the package.xml is the right one. Maybe for ROS 2 we need to restart from "zero" the version number.

```
  <version>1.4.0</version>
```

File: QUALITY_DECLARATION.md [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+
+This package requires that all changes occurr through a pull request.
+
+### Contributor Origin [2.ii]
+
+This package has a confirmation of contributor origin policy, which can be found in [CONTRIBUTING](./CONTRIBUTING.md).
+
+### Peer Review Policy [2.iii]
+
+Following the recommended guidelines for ROS Core packages, all pull request have at least 1 peer review.
```
Comment (brawner) [inline QUALITY_DECLARATION.md]:
```suggestion
Following the recommended guidelines for ROS Core packages, all pull requests must have at least 1 peer review.
```

File: QUALITY_DECLARATION.md [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+
+### Continuous Integration [2.iv]
+
+All pull request must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)
```
Comment (brawner) [inline QUALITY_DECLARATION.md]:
```suggestion
All pull requests must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)
```

File: QUALITY_DECLARATION.md [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+### Linters and Static Analysis [4.v]
+
+`class_loader` uses and passes all the standard linters and static analysis tools for a C++ package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis).
+
+## Dependencies [5]
+
+### Direct Runtime ROS Dependencies [5.i/5.ii]
+
+`rosidl_typesupport_cpp` does not have any runtime ROS 2 dependencies.
```
Comment (brawner) [inline QUALITY_DECLARATION.md]:
```suggestion
`class_loader` does not have any runtime ROS 2 dependencies.
```

File: QUALITY_DECLARATION.md [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+`class_loader` does not currently have performance tests.
+
+### Linters and Static Analysis [4.v]
+
+`class_loader` uses and passes all the standard linters and static analysis tools for a C++ package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis).
+
+## Dependencies [5]
+
+### Direct Runtime ROS Dependencies [5.i/5.ii]
```
Comment (brawner) [inline QUALITY_DECLARATION.md]:
```suggestion
### Direct Runtime ROS Dependencies [5.i]/[5.ii]
```

File: README.md [hunk 19]
```diff
@@ -0,0 +1,53 @@
+# class_loader
+
+The **class_loader** package is a ROS-independent package for loading plugins during runtime and the foundation of the higher level ROS `pluginlib` library. **class_loader** utilizes the host operating system's runtime loader to open runtime libraries (e.g. .so/.dll/dylib files), introspect the library for exported plugin classes, and allows users to instantiate objects of said exported classes without the explicit declaration (i.e. header file) for those classes.
+
+## class_loader VS pluginlib
+`class_loader` is used in the implementation of the higher-level ROS package `pluginlib` which is the encouraged method for loading plugins in the ROS ecosystem. You should use class_loader when creating plugins intended for non-ROS packages and pluginlib when exporting plugins to ROS packages.
+
+## Quality Declaration
+
+This package claims to be in the **Quality Level 1** category, see the [Quality Declaration](./QUALITY_DECLARATION.md) for more details.
```
Comment (brawner) [inline README.md]: Set to actual QL level

**Review by brawner — COMMENTED** [review]
Just a couple more nits

File: QUALITY_DECLARATION.md [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+
+## Version Policy [1]
+
+### Version Scheme [1.i]
+
+`class_loader` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#versioning), and is at or above a stable version.
+
+### Version Stability [1.ii]
+
+The current version is 1.4.0.
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: Like I mentioned, I don't think we want to have the specific version here unless we can automate it as this will get behind almost immediately when they bump versions and release it. I would just say something like, "It is at a stable version and currently above 1.0.0. See the [package.xml](package.xml) for it's current version."

File: QUALITY_DECLARATION.md [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+### Peer Review Policy [2.iii]
+
+Following the recommended guidelines for ROS Core packages, all pull requests must have at least 1 peer review.
+
+### Continuous Integration [2.iv]
+
+All pull requests must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)
+
+Currently nightly results can be seen here:
+* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/class_loader/)
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: Point to the ubuntu one instead (https://ci.ros2.org/job/nightly_linux_release/)

File: QUALITY_DECLARATION.md [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+* [linux-arm64_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/class_loader/)
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: This says arm64, but it points to nightly_linux_release, switch the link to arm64.

File: QUALITY_DECLARATION.md [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+### Documentation Policy [2.v]
+
+All pull requests must resolve related documentation changes before merging
+
+## Documentation [3]
+
+### Feature Documentation [3.i]
+
+There is no current ROS 2 feature list, but the feature list for ROS 1 can be found [here](http://wiki.ros.org/class_loader) and each item in the list links to the corresponding feature documentation.
+There is documentation for all of the features, and new features require documentation before being added.
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: > There is documentation for all of the features, and new features require documentation before being added.

This seems misleading if there is not yet a ROS 2 features list, and it would be impossible to enforce the second part. Maybe just remove this sentence for now.

File: QUALITY_DECLARATION.md [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+Most recent test results can be found [here](http://build.ros2.org/view/Epr/job/Epr__class_loader__ubuntu_bionic_amd64/lastBuild/testReport/class_loader/copyright_check_tests_only/)
+
+## Testing [4]
+
+### Feature Testing [4.i]
+
+Each feature in `class_loader` has corresponding tests which simulate typical usage, and they are located in the `test` directory.
+
+Currently nightly results can be seen here:
+* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/class_loader/)
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: Same comment as above

File: QUALITY_DECLARATION.md [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+The tests aim to cover both typical usage and corner cases, but are quantified by contributing to code coverage.
+
+### Coverage [4.iii]
+
+`class_loader` follows the recommendations for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#coverage), and opts to use branch coverage instead of line coverage.
+
+This includes:
+
+- tracking and reporting line coverage statistics
+- achieving and maintaining a reasonable branch line coverage (90-100%)
```
Comment (brawner) [inline QUALITY_DECLARATION.md:111]: Since 90% doesn't quite meet the ROS 2 core guidelines, I would state this as something more like, "This package does not yet meet the 95% coverage guideline, but it is currently above XY%

File: QUALITY_DECLARATION.md [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+- no lines are manually skipped in coverage calculations
+
+Changes are required to make a best effort to keep or increase coverage before being accepted, but decreases are allowed if properly justified and accepted by reviewers.
+
+Current coverage statistics can be viewed [here](https://ci.ros2.org/job/ci_linux_coverage/lastBuild/cobertura/).
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: Point to the specific build number (85 in this case) so that the link remains durable. Also, since this was a build of everything, I would also refine the link to point to the class_loader specific stuff so it's easier to verify the results. Also these two results suggest the coverage is more in the 80s. Is there something I'm missing?

https://ci.ros2.org/job/ci_linux_coverage/85/cobertura/src_ros_class_loader_include_class_loader/
https://ci.ros2.org/job/ci_linux_coverage/85/cobertura/src_ros_class_loader_src/

File: QUALITY_DECLARATION.md [hunk 27]
```diff
@@ -0,0 +1,199 @@
...
+
+It has one "buildtool" dependency, which do not affect the resulting quality of the package, because they do not contribute to the public library API.
+It also has several test dependencies, which do not affect the resulting quality of the package, because they are only used to build and run the test code.
+
+## Platform Support [6]
+
+`class_loader` supports all of the tier 1 platforms as described in [REP-2000](https://www.ros.org/reps/rep-2000.html#support-tiers), and tests each change against all of them.
+
+Currently nightly results can be seen here:
+* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/class_loader/)
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: Same comment as above

File: README.md (hunk 19 above)
Comment (brawner) [reply README.md]: This wasn't actually done. It's at 2 right?

File: QUALITY_DECLARATION.md [hunk 28]
```diff
@@ -0,0 +1,199 @@
...
+`class_loader` uses and passes all the standard linters and static analysis tools for a C++ package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis).
+
+## Dependencies [5]
+
+`class_loader` has run-time and build-time dependencies that need to be considered for this declaration.
+
+### Direct Runtime ROS Dependencies [5.i]/[5.ii]
+
+- console_bridge_vendor
+- rcpputils
```
Comment (chapulina) [inline QUALITY_DECLARATION.md:133]: `rcpputils` is [claiming level 4](https://github.com/ros2/rcpputils/pull/47/files#diff-92440392597dfc3b7adb67bc4ed481b3R5), so this package can't be at a lower level than its direct runtime dependencies.

That's why it's important we write the declaration for the lower level packages first.

**Comment by brawner** [comment]
Due to my own confusion, when linking to CI results we actually want to point to aarch64 jenkins jobs not arm64. [Rep-2000](https://www.ros.org/reps/rep-2000.html) says arm64, but ci_launcher runs ci for aarch64. Sorry for the misleading comments in this PR and your other one. 

The arm64 links you have do need to be fixed though.

File: QUALITY_DECLARATION.md (hunk 26 above)
Comment (ahcorde) [reply QUALITY_DECLARATION.md]: I use last build to know the current status. But I'm okey with a static value.

Fixed links

**Comment by ahcorde** [comment]
@brawner where can I find the arm64 nightly job? https://ci.ros2.org/view/nightly/

File: QUALITY_DECLARATION.md [hunk 29]
```diff
@@ -50,9 +50,11 @@ Following the recommended guidelines for ROS Core packages, all pull requests mu
 
 All pull requests must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)
 
+nightly_linux-/
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: Remove

File: QUALITY_DECLARATION.md [hunk 30]
```diff
@@ -50,9 +50,11 @@ Following the recommended guidelines for ROS Core packages, all pull requests mu
...
+
 Currently nightly results can be seen here:
-* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/class_loader/)
-* [linux-arm64_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/class_loader/)
+* [linux-aarch64_release](https://ci.ros2.org/job/nightly_linux-aarch64_release/lastBuild/testReport/class_loader/)
+* [linux-arm64_release](https://ci.ros2.org/job/nightly_linux_release/lastBuild/testReport/class_loader/)
```
Comment (brawner) [inline QUALITY_DECLARATION.md]: Rename this link because it's not arm64. Fix this one and the several below.

**Review by brawner — APPROVED** [review]
Fix these two things and this one is good to go.
